### PR TITLE
Fix nodeport service compat with default-deny fw

### PR DESCRIPTION
**Pull request by marun**
This PR is a first pass at fixing #18802.  It adds explicit iptables rules allowing NodePort service traffic originating from outside a node to avoid the default-deny rules configured on non-cloud/vm deployments of fedora/centos/rhel.  

As per a conversation on #kubernetes-dev with @thockin, the goal was to:
- add a chain to the filter table 
- add rules to that chain per protocol that accepted traffic on the range of active proxy ports

I assumed that it was necessary to maintain availability of services when modifying the rule, and this implied adding a rule with an updated range before removing the rule with the previous range.  Since multiple iptables operations cannot be executed atomically, however, there exists the possibility that a new rule would be added and removal of the old rule would fail.  Subsequently removing an old rule is complicated by the current implementation for rule deletion, which requires the exact rule to target for deletion.  Since the rules in question include the port range for a given protocol at a point in time, attempting deletion of 'orphaned' rules would require keeping track of any protocol and port range pairs that were known to have been orphaned and repeatedly attempting their deletion.  I'm not sure this would be worth the trouble, though, and think the following alternatives would be preferable:
- implement chain-local orphan deletion.  When updating the rules, delete all rules in the chain that don't correspond to the current protocol/port-range pairs.  This approach adds some complexity in iptables manipulation for the sake of rule simplicity.
- add a rule per-service rather than using a rule per protocol with ranges.  This would be much simpler to implement, but would result in an additional rule per proxy port.

I've avoided adding unit tests for the new code just yet since the implementation is likely to change.  Ensuring e2ed coverage for the change will be complicated by the distro-specific nature of the problem, though existing tests will still detect regressions.

### CODE REVIEW

**Comment by knobunc** [comment]
I don't like this because with these rules, we will accept all external traffic that goes directly to a proxy port.  So an attacker could portscan the box and then hit services that were not exposed by a nodeport by talking directly to the random addresses.

Also, there's no guarantee that the proxier is the only thing requesting ports on the box, so the min and max of the range may expose sockets for unrelated things.

I think we should make a bitmap:port ipset to allow just the right ports.  Then you can install the ipset change atomically http://ipset.netfilter.org/tips.html

**Comment by marun** [comment]
As per @knobunc, the current commit neglected to differentiate between nodeport and non-nodeport proxy ports.  Allowing non-nodeport proxy ports has the potential to allow off-node traffic to target services that should be private to the cluster.  Since not all proxy ports should be treated equally, I think  rule-per-port is the best option here.  ipset would definitely be preferable, but I think that it is an optimization that should be tackled separately from this fix.

**Comment by marun** [comment]
The latest commit adds a rule allowing non-local traffic to each NodePort service proxy port.

**Comment by marun** [comment]
The unit tests have limited ability to validate iptables manipulation.  The existing e2e service tests won't be validating compatibility with default deny without test setup changes (e.g. explicitly installing iptables-services package on fedora).  Should test setup changes be included in this proposal, and for specific or  all distros and providers?

**Comment by knobunc** [comment]
LGTM

File: pkg/proxy/userspace/proxier.go [hunk 1]
```diff
@@ -972,6 +1015,13 @@ func (proxier *Proxier) iptablesHostNodePortArgs(nodePort int, protocol api.Prot
 	return args
 }
 
+// Build a slice of iptables args for an from-non-local public-port rule.
+func (proxier *Proxier) iptablesNonLocalNodePortArgs(nodePort int, protocol api.Protocol, proxyIP net.IP, proxyPort int, service proxy.ServicePortName) []string {
+	args := iptablesCommonPortalArgs(nil, false, false, proxyPort, protocol, service)
+	args = append(args, "-m", "state", "--state", "NEW", "-j", "ACCEPT")
```
Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: please add a `"-m", "comment", "--comment", service.String()` - it makes for easier grep

**Comment by thockin** [comment]
Your comments are way scarier than this PR warrants.

What about the iptables proxier?  It seems that you and @maclof (https://github.com/maclof/kubernetes/commit/d394dfb96ebe73cc4c02835813dd5e3effef4285) could collab - where he puts a REJECT, you need an ACCEPT, it seems.

**Comment by thockin** [comment]
re: e2e, this is a tough case.  Can you leave a TODO in the Services e2e that we need an iptables test on a more locked-down installation

**Comment by marun** [comment]
@thockin I'm confused, did you mean to close this PR?

**Comment by eparis** [comment]
I re-openned, tim can re-close if he wants   :)

**Comment by thockin** [comment]
Sorry, no, that is s not what I intended :)

**Comment by marun** [comment]
@thockin I don't think the iptables proxy will need a fix to be compatible with a default deny firewall.  Inbound NodePort traffic will be rewritten to target the pod ip/port which must already be allowed to ensure intra-node pod connectivity.

**Comment by marun** [comment]
I had tested this on openshift before it switched to the iptables proxy.  I'm been trying to create a local vagrant+virtualbox cluster but have been seeing deployment failures related to flannel.

File: pkg/proxy/userspace/proxier.go (hunk 1 above)
Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: I think this and a TODO for e2e are the only things blocking LGTM

Comment (thockin-cc) [inline pkg/proxy/userspace/proxier.go]: Any chance this PR will get some attention before the code freeze next week?

Comment (marun) [inline pkg/proxy/userspace/proxier.go]: Sorry for the delay, I was off TH-F last week.  I'll get this done today.

Comment (thockin) [inline pkg/proxy/userspace/proxier.go]: Awesome.  ping me when ready.

**Comment by marun** [comment]
@thockin When I installed iptables-services on the nodes of a kube dev cluster, it broke heapster (pod would endlessly restart).  If/when e2e testing is pursued for validating compatibility of userspace proxy whitelisting with a default deny firewall, it will likely be necessary to ensure that heapster traffic is also whitelisted.

Otherwise, I think this is ready to go in.

**Comment by marun** [comment]
Looks like e2e setup failed.  How to re-trigger the job?

**Comment by eparis** [comment]
step #1 is find or file an issue.
step #2 is tell k8s-bot test this github issue: #NUMBER

**Comment by marun** [comment]
@k8s-bot test this issue: #20306

**Comment by marun** [comment]
@eparis Anything more to do to allow merge robot to work?  The check summary seems to indicate that this PR is not in the submit queue due to ci tests not being green.

**Comment by eparis** [comment]
you were just too fast.  the poor little submit queue polls every 10 minutes.

**Pull request merged**